accel/tcg/log_instr.c

```diff
@@ -0,0 +1,902 @@
+                 pc, cpu_get_asid(env));
+    else
+        qemu_log("Requested instruction logging @ " TARGET_FMT_lx
+                 " ASID %u\n", pc, cpu_get_asid(env));
+}
+
+/*
+ * Emit text tracing stop event.
+ */
+static void emit_text_stop(CPUArchState *env, target_ulong pc)
+{
+    cpu_log_instr_info_t *log = cpu_get_log_instr_state(env);
+
+    if (qemu_loglevel_mask(CPU_LOG_USER_ONLY) && log->user_mode_tracing)
+        qemu_log("User-mode only tracing disabled at " TARGET_FMT_lx
+                 ", ASID %u\n", pc, cpu_get_asid(env));
+    else
+        qemu_log("Disabled instruction logging @ " TARGET_FMT_lx
+                 " ASID %u\n", pc, cpu_get_asid(env));
+}
+
+/* CHERI trace V3 format emitters */
+
+/*
+ * Emit cvtrace trace trace header. This is a magic byte + string
+ */
+static void emit_cvtrace_header(CPUArchState *env)
+{
+    FILE *logfile = qemu_log_lock();
+    char buffer[sizeof(cheri_trace_entry_t)];
+
+    buffer[0] = CTE_QEMU_VERSION;
+    g_strlcpy(buffer + 1, CTE_QEMU_MAGIC, sizeof(buffer) - 2);
+    fwrite(buffer, sizeof(buffer), 1, logfile);
+    qemu_log_unlock(logfile);
+}
+
+/*
+ * Emit cvtrace trace entry.
+ */
+static void emit_cvtrace_entry(CPUArchState *env, cpu_log_instr_info_t *log)
+{
+    FILE *logfile;
+    cheri_trace_entry_t entry;
+    static uint16_t cycles = 0; // TODO(am2419): this should be a per-cpu counter.
+
+    entry.entry_type = CTE_NO_REG;
+    entry.thread = (uint8_t)env_cpu(env)->cpu_index;
+    entry.asid = (uint8_t)log->asid;
+    entry.pc = cpu_to_host64(log->pc);
+    entry.cycles = cpu_to_host16(cycles++);
+    entry.inst = *((uint32_t *)&log->insn_bytes[0]);
+    /* entry.inst = log->opcode; // TODO(am2419): opcode, how to pick it up? */
+    switch (log->flags & LI_FLAG_INTR_MASK) {
+    case LI_FLAG_INTR_TRAP:
+        entry.exception = (uint8_t)log->intr_code;
+    case LI_FLAG_INTR_ASYNC:
+        entry.exception = 0; // TODO(am2419): this is very MIPS-specific.
+    default:
+        entry.exception = CTE_EXCEPTION_NONE;
+    }
+
+    if (log->regs->len) {
+        log_reginfo_t *rinfo = &g_array_index(log->regs, log_reginfo_t, 0);
+#ifndef TARGET_CHERI
+        log_assert(!reginfo_is_cap(rinfo) && "Capability register access "
+                   "without CHERI support");
+#else
+        if (reginfo_is_cap(rinfo)) {
+            cap_register_t intcap;
+            cap_register_t *cr = &rinfo->cap;
+
+            if (!reginfo_has_cap(rinfo)) {
+                // cvtrace expects a null capability in the integer case
+                cr = null_capability(&intcap);
+            }
+            uint64_t metadata = (
+                ((uint64_t)cr->cr_tag << 63) | (cap_get_otype(cr) << 32) |
+                (COMBINED_PERMS_VALUE(cr) << 1) |
+                (uint64_t)(cap_is_unsealed(cr) ? 0 : 1));
+
+            entry.entry_type = CTE_CAP;
+            entry.val2 = cpu_to_host64(metadata);
+            entry.val3 = cpu_to_host64(cap_get_cursor(cr));
```

Comment:
cap_get_cursor returns a host endian value. If the cvtrace format is big endian you should use cpu_to_be64.

Reply:
Right, I got it backwards.